experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,732 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	dev := &Dev{
+		spiDev:           spiDev,
+		operationTimeout: 30 * time.Second,
+	}
+
+	dev.resetPin = resetPin
+	dev.resetPin.Out(gpio.High)
+
+	dev.irqPin = irqPin
+	dev.irqPin.In(gpio.PullUp, gpio.FallingEdge)
+
+	err = dev.Init()
+
+	return dev, err
+}
+
+// SetOperationtimeout updates the device timeout for card operations.
+// Effectively that sets the maximum time the RFID device will wait for IRQ from the proximity card detection.
+//
+//	timeout the duration to wait for IRQ strobe.
+func (r *Dev) SetOperationtimeout(timeout time.Duration) {
+	r.operationTimeout = timeout
+}
+
+var sequenceCommands = struct {
+	init     [][]byte
+	waitInit [][]byte
+	waitLoop [][]byte
+}{
+	init: [][]byte{
+		{commands.TModeReg, 0x8D},
+		{commands.TPrescalerReg, 0x3E},
+		{commands.TReloadRegL, 30},
+		{commands.TReloadRegH, 0},
+		{commands.TxAutoReg, 0x40},
+		{commands.ModeReg, 0x3D},
+	},
+	waitInit: [][]byte{
+		{commands.CommIrqReg, 0x00},
+		{commands.CommIEnReg, 0xA0},
+	},
+	waitLoop: [][]byte{
+		{commands.FIFODataReg, 0x26},
+		{commands.CommandReg, 0x0C},
+		{commands.BitFramingReg, 0x87},
+	},
+}
+
+func (r *Dev) writeCommandSequence(commands [][]byte) error {
+	for _, cmdData := range commands {
+		if err := r.devWrite(int(cmdData[0]), cmdData[1]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() error {
+	if err := r.Reset(); err != nil {
+		return err
+	}
+	if err := r.writeCommandSequence(sequenceCommands.init); err != nil {
+		return err
+	}
+	return r.SetAntenna(true)
+}
+
+func (r *Dev) devWrite(address int, data byte) error {
+	newData := []byte{(byte(address) << 1) & 0x7E, data}
+	return r.spiDev.Tx(newData, nil)
+}
+
+func (r *Dev) devRead(address int) (byte, error) {
+	data := []byte{((byte(address) << 1) & 0x7E) | 0x80, 0}
+	out := make([]byte, len(data))
+	if err := r.spiDev.Tx(data, out); err != nil {
+		return 0, err
+	}
+	return out[1], nil
+}
+
+func (r *Dev) setBitmask(address, mask int) error {
+	current, err := r.devRead(address)
+	if err != nil {
+		return err
+	}
+	return r.devWrite(address, current|byte(mask))
+}
+
+func (r *Dev) clearBitmask(address, mask int) error {
+	current, err := r.devRead(address)
+	if err != nil {
+		return err
+	}
+	return r.devWrite(address, current&^byte(mask))
+}
+
+// Reset resets the RFID chip to initial state.
+func (r *Dev) Reset() error {
+	return r.devWrite(commands.CommandReg, commands.PCD_RESETPHASE)
+}
+
+// Halt soft-stops the chip - PowerDown bit set, command IDLE
+func (r *Dev) Halt() error {
+	return r.devWrite(commands.CommandReg, 16)
+}
+
+// SetAntenna configures the antenna state, on/off.
+func (r *Dev) SetAntenna(state bool) error {
+	if state {
+		current, err := r.devRead(commands.TxControlReg)
+		if err != nil {
+			return err
+		}
+		if current&0x03 != 0 {
+			return errors.New("Can not set the bitmask for antenna")
```

Comment:
Eventually you'll want to add a wrapf() function that wraps all errors with a "mfrc522:" prefix, so when the error is surfaced to the user the user can figure out where the error comes from. You can grep for "wrapf" in the code base for examples.

Reply:
okay, makes sense